apiserver/keymanager/keymanager.go

```diff
@@ -54,48 +54,70 @@ var _ KeyManager = (*KeyManagerAPI)(nil)
 // NewKeyManagerAPI creates a new server-side keyupdater API end point.
 func NewKeyManagerAPI(st *state.State, resources facade.Resources, authorizer facade.Authorizer) (*KeyManagerAPI, error) {
 	// Only clients and environment managers can access the key manager service.
-	if !authorizer.AuthClient() && !authorizer.AuthModelManager() {
+	if !authorizer.AuthClient() && !authorizer.AuthController() {
 		return nil, common.ErrPerm
 	}
-	env, err := st.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	// For gccgo interface comparisons, we need a Tag.
-	owner := names.Tag(env.Owner())
-	// TODO(wallyworld) - replace stub with real canRead function
-	// For now, only admins can read authorised ssh keys.
-	canRead := func(user string) bool {
-		// Are we a machine agent operating as the system identity?
-		if user == config.JujuSystemKey {
-			_, ismachinetag := authorizer.GetAuthTag().(names.MachineTag)
-			return ismachinetag
-		}
-		return authorizer.GetAuthTag() == owner
-	}
-	// TODO(wallyworld) - replace stub with real canWrite function
-	// For now, only admins can write authorised ssh keys for users.
-	// Machine agents can write the juju-system-key.
-	canWrite := func(user string) bool {
-		// Are we a machine agent writing the Juju system key.
-		if user == config.JujuSystemKey {
-			_, ismachinetag := authorizer.GetAuthTag().(names.MachineTag)
-			return ismachinetag
-		}
-		// No point looking to see if the user exists as we are not
-		// yet storing keys on the user.
-		return authorizer.GetAuthTag() == owner
+	var apiUser *names.UserTag
+	if authorizer.AuthClient() {
+		u, _ := authorizer.GetAuthTag().(names.UserTag)
+		apiUser = &u
 	}
 	return &KeyManagerAPI{
 		state:      st,
 		resources:  resources,
 		authorizer: authorizer,
-		canRead:    canRead,
-		canWrite:   canWrite,
+		apiUser:    apiUser,
 		check:      common.NewBlockChecker(st),
 	}, nil
 }
 
+func (api *KeyManagerAPI) checkCanRead(sshUser string) error {
+	if api.checkCanWrite(sshUser) == nil {
+		return nil
+	}
+	if api.apiUser == nil {
+		return common.ErrPerm
+	}
+	ok, err := common.HasPermission(
+		api.state.UserAccess,
+		*api.apiUser,
+		permission.ReadAccess,
+		api.state.ControllerTag(),
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !ok {
+		return common.ErrPerm
+	}
+	return nil
+}
+
+func (api *KeyManagerAPI) checkCanWrite(sshUser string) error {
+	// Are we a machine agent writing the Juju system key.
+	if api.apiUser == nil {
+		if sshUser == config.JujuSystemKey {
+			_, ismachinetag := api.authorizer.GetAuthTag().(names.MachineTag)
```

Comment:
> It's not such much not wanting to make it better - but not wanting to mess with the existing behaviour. If the ssh key param was being checked before (even after we knew it was a machine agent connecting), I didn't want to change that now. Otherwise a machine agent could change something other than the system key.

Maybe there's some misunderstanding, because I don't understand your response. I didn't mean we should stop checking the ssh key param. I meant that the "ismachinetag" bits are redundant, since we already know the authorized entity is a controller machine agent.

> I though it would be ok for admins to be able to modify the system key - they have admin right? But maybe that's not a good idea? I could check that for users, it's not the system key

I would err on the side of caution and leave it as it was: have it return ErrPerm if a user attempts to modify (or read!) the system identity key.